daemon/daemon_unix.go

```diff
@@ -1364,12 +1364,25 @@ func (daemon *Daemon) registerLinks(container *container.Container, hostConfig *
 // conditionalMountOnStart is a platform specific helper function during the
 // container start to call mount.
 func (daemon *Daemon) conditionalMountOnStart(container *container.Container) error {
-	return daemon.Mount(container)
+	if err := daemon.Mount(container); err != nil {
+		return err
+	}
+	// Change the container rootfs mount propagation mode to prevent any
+	// temporary mounts we create outside the container (using
+	// (*Daemon).mountVolumes) from propagating into the container's mount
+	// namespace. Unconditionally bind-mount over the rootfs so we can
+	// unconditionally unmount it on cleanup without a chance of unmounting
+	// any mounts managed by the graph driver.
+	rootfs := container.BaseFS.Path()
+	return mount.Mount(rootfs, rootfs, "", "bind,private")
```

Comment:
Instead of making each container rootfs private, we could solve the issue in `daemon.setupDaemonRootPropagation` by ensuring the propagation of `/var/lib/docker` is slave or private rather than the current behaviour of ensuring it is shared or slave. WDYT?

Reply:
That might not be a great idea, come to think of it. If `/var/lib/docker` had slave or private propagation and a container is started with `/var/lib/docker` as a volume, any rootfs mounts for other containers would get copied into the container namespace with a propagation mode which would prevent unmounts of the other containers' rootfs'es from propagating into that container? Should the daemon be forcing the propagation mode of `/var/lib/docker` to shared(+slave)?